### PR TITLE
OpenAI - update semantic conventions to the latest version, write logs based events instead of span events

### PR DESCRIPTION
**Pull request by lmolkova**
# Description

- Update OpenAI semantic conventions to the 1.28.0 
  - changes span events to log based events 
  - adds a few new openai-specific attributes
  - adds opt-in flag for event contents
- Add tests

TODOs:
- [x] we need opt-in flag for contents 
- [x] changelog

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [ ] Documentation has been updated

### CODE REVIEW

File: instrumentation/opentelemetry-instrumentation-openai-v2/tests/conftest.py
Comment (lmolkova) [inline instrumentation/opentelemetry-instrumentation-openai-v2/tests/conftest.py:84]: TODO: This was running before the test, will need to figure out how to do clean up right

File: CONTRIBUTING.md
Comment (xrmx) [inline CONTRIBUTING.md:230]: I'm not sure we want to keep this up to date :sweat_smile:

File: instrumentation/opentelemetry-instrumentation-openai-v2/tests/conftest.py
Comment (xrmx) [reply instrumentation/opentelemetry-instrumentation-openai-v2/tests/conftest.py:84]: How does it fail?

Comment (xrmx) [reply instrumentation/opentelemetry-instrumentation-openai-v2/tests/conftest.py:84]: To me the correct way to handle this would be to do it per test:

```
@pytest.fixture(autouse=True)
def instrument():
    instrumentor = OpenAIInstrumentor()
    instrumentor.instrument()

    yield instrumentor

    instrumentor.uninstrument()
```

File: instrumentation/opentelemetry-instrumentation-openai-v2/tests/test_chat_completions.py
Comment (xrmx) [inline instrumentation/opentelemetry-instrumentation-openai-v2/tests/test_chat_completions.py:634]:
```suggestion
    assert dict(log.log_record.body) == expected_content
```

Have you tried something like this? It's cheaper and the comparison is specialized so key order does not matter and you get the key that is different highlighted already

File: instrumentation/opentelemetry-instrumentation-openai-v2/tests/test_chat_completions.py
Comment (xrmx) [inline instrumentation/opentelemetry-instrumentation-openai-v2/tests/test_chat_completions.py:546]: Since these are recorded shouldn't we just hardcode the numbers later instead? this way we are copying the implementation that may hide any issue.

File: instrumentation/opentelemetry-instrumentation-openai-v2/tests/conftest.py
Comment (lmolkova) [reply instrumentation/opentelemetry-instrumentation-openai-v2/tests/conftest.py:84]: thanks! will try!  how it failed - the instrumentation was disabled (since uninstrument was called before test started),

File: CONTRIBUTING.md
Comment (lmolkova) [reply CONTRIBUTING.md:230]: I need a few features from the not-yet-released core (semconv 1.28.0 and event SDK for tests). I think I can work around both needs by hardcoding attribute names and implementing custom event SDK in tests.

but out of curiosity, how do we normally handle things like this - wait for the core repo to release?

A related question - it seems we last released core two months ago - is a new release planned any time soon or should I do the workarounds?

File: instrumentation/opentelemetry-instrumentation-openai-v2/tests/test_chat_completions.py
Comment (lmolkova) [reply instrumentation/opentelemetry-instrumentation-openai-v2/tests/test_chat_completions.py:634]: thanks, will try!

File: CONTRIBUTING.md
Comment (xrmx) [reply CONTRIBUTING.md:230]: By default it should use main so you won't need to touch anything usually

File: instrumentation/opentelemetry-instrumentation-openai-v2/tests/test_chat_completions.py
Comment (lmolkova) [reply instrumentation/opentelemetry-instrumentation-openai-v2/tests/test_chat_completions.py:546]: if we hardcode everything, then it'd become a problem if we ever need to re-record. From my experience working with test recordings the need comes every once in a while.

File: CONTRIBUTING.md
Comment (lzchen) [reply CONTRIBUTING.md:230]: > but out of curiosity, how do we normally handle things like this - wait for the core repo to release?

The SHA points to a commit in the core repo (usually unreleased when you change it) to pull dev changes from the other repository that is needed for the current PR. As @xrmx has mentioned, USUALLY you can set it to `main` to pick up the latest changes from core. In the off chance that you make two simultaneous PRs in both core and contrib that need changes from each other, this is when we usually set the SHA to specific commits of the respective PRs.

> A related question - it seems we last released core two months ago - is a new release planned any time soon or should I do the workarounds?

We will probably do a release soon once this gets merged.

File: instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/__init__.py
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/__init__.py:70]: Should this be `event_logger_provider`?

File: instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/patch.py
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/patch.py:87]: Do we need this check if it already exists in `chat_completions_create`

File: instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py:166]: > Conditionally Required: if available and if not equal to assistant

Does that mean if it IS equal to "assistant", we do not populate this attribute? From the sem conv, it's not obvious to me that "assistant" is the default value if nothing is there.

File: instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py:171]: NIT: Put a TODO for opt-in functionality?

File: instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py:177]: Nit: Should we have a collective place where we can refer to different event names (constants.py?) or might be good to add to sem conv.

File: instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py:72]: Could these possibly be `None`? Should we skip a call if any of these fields don't exist?

File: instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py:86]: From the definition of [Function object](https://github.com/open-telemetry/semantic-conventions/blob/v1.28.0/docs/gen-ai/gen-ai-events.md#function-object), it seems like `arguments` can be `AnyValue` so `replace` would probably error out in those cases. Perhaps we simply leave the value as is?

File: instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py:127]: Just curious, are there `role` values that are not currently defined in sem conv? (system, user, assistant, tool, etc)?

File: instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/patch.py
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/patch.py:149]: I'm not too familiar with the concept of `Stream`. How does it conceptually map to a `Choice` or number of `Choice`s?

File: CONTRIBUTING.md
Comment (xrmx) [reply CONTRIBUTING.md:230]: > We will probably do a release soon once this gets merged.

We can release earlier since this instrumentation will be released separately on its own

File: instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py
Comment (lmolkova) [reply instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py:166]: image

https://github.com/open-telemetry/semantic-conventions/blob/v1.28.0/docs/gen-ai/gen-ai-events.md#assistant-event

File: instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py
Comment (lmolkova) [reply instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py:86]: `arguments` is a string in openai REST api and client.  Semconv leave it undefined for the future. I can check if it's a sting - I don't want to have new lines in the logs, so replacing is kinda important

File: instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py
Comment (lmolkova) [reply instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py:127]: not that I'm aware of, but I believe users can pass any role and it may work in some edge cases or with some non-OpenAI backends (a lot of backends are Open-AI compatible on the REST level and this instrumentation can be used with any of them).

File: instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/patch.py
Comment (lmolkova) [reply instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/patch.py:149]: You can ask model to return a completion (one or more choices) using regular HTTP response body or you can ask it to stream tokens one-by-one - https://platform.openai.com/docs/api-reference/making-requests - in which case model returns the HTTP response and then writes tokens to the response stream using  SSE events. Check out  test recordings to see what happens there with and without streaming.

I.e. streaming affects how the response is received, but not the contents (choices).

File: instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py
Comment (lmolkova) [reply instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py]: good catch, removed default

File: instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py
Comment (lmolkova) [reply instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py]: added check for none, we should not really skip anything - let's just translate what we have into event

File: instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py:39]: NIt: Any truthy value will return true (any non-empty string would return true). Do we want that or do we want explicitly "true" string like we have for other [env vars](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-logging/src/opentelemetry/instrumentation/logging/__init__.py#L132). Not a strong opinion though.

**Review by lzchen — APPROVED** [review]
LGTM!

File: instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py
Comment (lmolkova) [reply instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py]: great catch, thanks!

**Review by karthikscale3 — APPROVED** [review]
looking good. nice work @lmolkova

**Comment by alizenhom** [comment]
looking good, thanks for doing this @lmolkova

**Review by alizenhom — APPROVED** [review]

File: instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py
Comment (xrmx) [inline instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py:40]:
```suggestion
        OTEL_INSTRUMENTATION_OPENAI_CAPTURE_MESSAGE_CONTENT, "false"
    )

    return capture_content.lower() == "true"
```

File: instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py
Comment (xrmx) [inline instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py:128]: These should be set by default at emit() time, are you adding them explicitly because they do not match?

**Review by xrmx — APPROVED** [review]

File: instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py
Comment (lmolkova) [reply instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py]: great catch! Changed to implicit context, except the streaming case where the span associated with the stream is not current after the original method returns.

**Pull request merged**
